go/citogo/main.go

```diff
@@ -148,106 +149,98 @@ func (r *runner) flushTestLogsToTemp(logName string, log bytes.Buffer) (string,
-	hook := os.Getenv("CITOGO_FLAKE_WEBHOOK")
-	if hook == "" {
+func (r *runner) report(result TestResult) {
+	b, err := json.Marshal(result)
+	if err != nil {
+		logError("error marshalling result: %s", err.Error())
 		return
 	}
 
-	r.doHook(hook, test, logs, "❄️")
-}
-
-func (r *runner) doHook(hook string, test string, logs string, emoji string) {
-	hook += url.QueryEscape(fmt.Sprintf("%s _client_ %s-%s %s *%s* %s [%s]", emoji, r.opts.Branch, r.opts.BuildID, r.opts.Prefix, test, logs, r.opts.BuildURL))
-	_, err := http.Get(hook)
+	err = lambdaInvoke("report-citogo", b)
 	if err != nil {
 		logError("error reporting flake: %s", err.Error())
 	}
 }
 
-type outcome string
+type Outcome string
 
 const (
-	success outcome = "success"
-	flake   outcome = "flake"
-	fail    outcome = "fail"
+	OutcomeSuccess Outcome = "success"
+	OutcomeFlake   Outcome = "flake"
+	OutcomeFail    Outcome = "fail"
 )
 
-func (o outcome) abbrv() string {
+func (o Outcome) Abbrv() string {
 	switch o {
-	case success:
+	case OutcomeSuccess:
 		return "PASS"
-	case flake:
+	case OutcomeFlake:
 		return "FLK?"
-	case fail:
+	case OutcomeFail:
 		return "FAIL"
 	default:
 		return "????"
 	}
 }
 
-func (r *runner) reportTestOutcome(outcome outcome, test string, where string) {
-	fmt.Printf("%s: %s", outcome.abbrv(), test)
-	if where != "" {
-		fmt.Printf(" %s", where)
+func (r *runner) runTest(test string) error {
+	outcome, where, err := r.runTestOnce(test, false /* isRerun */)
+	if err != nil {
+		return err
 	}
-	fmt.Printf("\n")
-
-	if outcome != fail || r.opts.Branch != "master" {
-		return
+	if outcome == OutcomeSuccess {
+		return nil
 	}
-	hook := os.Getenv("CITOGO_MASTER_FAIL_WEBHOOK")
-	if hook == "" {
-		return
+	if len(r.flakes) >= r.opts.Flakes {
+		return errTestFailed
 	}
-	r.doHook(hook, test, where, "🥴")
-}
-
-func (r *runner) runTest(test string) error {
-	canRerun := len(r.flakes) < r.opts.Flakes
-	logs, err := r.runTestOnce(test, canRerun, false)
-	if err == errTestFailed && canRerun {
-		_, err = r.runTestOnce(test, false, true)
-		if err == nil {
-			r.reportFlake(test, logs)
-			r.flakes = append(r.flakes, test)
-		}
+	outcome2, _, err2 := r.runTestOnce(test, true /* isRerun */)
+	if err2 != nil {
+		return err2
 	}
-	return err
+	switch outcome2 {
+	case OutcomeFail:
+		return errTestFailed
+	case OutcomeSuccess:
+		fmt.Printf("FLK: %s\n", test)
+		r.report(TestResult{Outcome: OutcomeFlake, TestName: test, Where: where, Branch: r.opts.Branch})
+		r.flakes = append(r.flakes, test)
+	}
+	return nil
 }
 
 var errTestFailed = errors.New("test failed")
 
-func (r *runner) runTestOnce(test string, canRerun bool, isRerun bool) (string, error) {
+// runTestOnce only returns an error if there was a problem with the test
+// harness code; it does not return an error if the test failed.
+func (r *runner) runTestOnce(test string, isRerun bool) (outcome Outcome, where string, err error) {
+	defer func() {
+		fmt.Printf("%s: %s %s\n", outcome.Abbrv(), test, where)
```

Comment:
this now doesn't work as before. It used to report `FLK?` if you could rerun it and it failed. Please restore the previous behavior.

Reply:
Done